src/main.c

```diff
@@ -73,6 +74,62 @@ static int set_all_thread_affinity(void)
     return 0;
 }
 
+static void dpvs_usage(const char *prgname)
+{
+    printf("\nUsage: %s ", prgname);
+    printf("DPVS application options:\n"
+            "   -v  version     display DPVS version info\n"
+            "   -h  help        display DPVS help info\n"
+    );
+}
+
+static int parse_app_args(int argc, char **argv)
+{
+    const char *short_options = "vh";
+    char *prgname = argv[0];
+    int c, ret = -1;
+
+    const int old_optind = optind;
+    const int old_optopt = optopt;
+    char * const old_optarg = optarg;
+
+    struct option long_options[] = {
+        {"version", 0, NULL, 'v'},
+        {"help", 0, NULL, 'h'},
+        {NULL, 0, 0, 0}
+    };
+
+    optind = 1;
+
+    while((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
+        switch(c) {
+            case 'v':
+                fprintf(stderr, "dpvs version: %s, build on %s\n", DPVS_VERSION, DPVS_BUILD_DATE);
+                exit(EXIT_SUCCESS);
+            case 'h':
+                dpvs_usage(prgname);
+                exit(EXIT_SUCCESS);
+            case '?':
+                dpvs_usage(prgname);
+                exit(EXIT_FAILURE);
+            default:
+                break;
+        }
+    }
+
+    if (optind > 0)
+        argv[optind-1] = prgname;
+
+    ret = optind - 1;
+
+    /* restore getopt lib */
+    optind = old_optind;
```

Comment:
If some argument are consumed, is old index still correct ?

Reply:
Yes. old_optind won't change after that. And all eal arguments parse function will reset optind to 1 as double assurance.